src/server.rs

```diff
@@ -165,7 +168,17 @@ impl UdpServer {
 
     /// Gossip to one other UDP server.
     async fn gossip(&self, addr: impl Into<String>) -> anyhow::Result<()> {
-        let syn = self.scuttlebutt.lock().await.create_syn_message();
+        let mut scuttlebutt = self.scuttlebutt.lock().await;
+        let heartbeat = scuttlebutt
+            .self_node_state()
+            .get(HEARTBEAT_KEY)
+            .and_then(|heartbeat| str::parse(heartbeat).ok())
+            .unwrap_or(0);
+        scuttlebutt
+            .self_node_state()
```

Comment:
You called self_node_state twice.
That thing takes a lock. That's a race condition here.
If you call gossip concurrently twice, the value for the heartbeat is not well-defined.

Reply:
I'm probably missing something, but the lock is held across both get and set of `self_node_state`. I don't see any locks in the model/lib so I'd assume you're talking about the `scuttlebutt` lock here?

In that scenario this shouldn't be any different than `let hb = scuttlebutt.hb + 1; scuttlebutt.hb = hb;` unless there's some chance for async Rust to put tasks into the background without any `.await` points in between (which I don't think it is? Admittedly not particularly experienced with async Rust).

Reply:
Ah yes my bad! I thought self node state was taking a lock and returning a mutexguard. Sorry I confused myself.